aiida/engine/daemon/execmanager.py

```diff
@@ -241,16 +241,22 @@ def find_data_node(inputs, uuid):
                 raise IOError('It is not possible to create a symlink between two different machines for '
                               'calculation {}'.format(node.pk))
 
-    # Before copying the files in the sandbox `folder`, first remove those that are flagged in `provenance_exclude_list`
     provenance_exclude_list = calc_info.provenance_exclude_list or []
-    for filepath in provenance_exclude_list:
-        logger.debug('`{}` found in `provenance_exclude_list` so will not be stored in the repository'.format(filepath))
-        folder.remove_path(filepath)
-
-    # Move the remaining contents of the sandbox `folder` to the repository of the `CalcJobNode` instance. Note that we
-    # use `force=True` such that if this was already called but the `upload_calculation` failed and was rescheduled, we
-    # simply overwrite the existing files of the previous attempt.
-    node.put_object_from_tree(folder.abspath, force=True)
+
+    # Loop recursively over content of the sandbox folder copying all that are not in `provenance_exclude_list`. Note
+    # that directories are not created explicitly. The `node.put_object_from_filelike` call will create intermediate
+    # directories for nested files automatically when needed. This means though that empty folders in the sandbox or
+    # folders that would be empty when considering the `provenance_exclude_list` will *not* be copied to the repo. The
+    # advantage of this explicit copying instead of deleting the files from `provenance_exclude_list` from the sandbox
+    # first before moving the entire remaining content to the node's repository, is that in this way we are guaranteed
+    # not to accidentally move files to the repository that should not go there at all cost.
+    for root, dirnames, filenames in os.walk(folder.abspath):
+        for filename in filenames:
+            filepath = os.path.join(root, filename)
+            relpath = os.path.relpath(filepath, folder.abspath)
+            if relpath not in provenance_exclude_list:
```

Comment:
Is there a way that `relpath` and `provenance_exclude_list` can refer to the same file, but not contain the same string? Except of course if the plugin supply something that is not correct for `provenance_exclude_list`?

Reply:
If the `provenance_exclude_list` is defined correctly, then no, this should work. The `provenance_exclude_list` should contain file paths relative to the base path of the `folder` sandbox. Since the files therein are created also with relative filepaths in the `prepare_for_submission` where the `provenance_exclude_list` is created, there is a one-to-one correspondence. So I think the risk for mistakes should be minimal.
Example would be:
```
def prepare_for_submission(self, folder):
    provenance_exclude_list = []
    for element, pseudo in self.inputs.pseudos.items():
        with pseudo.open(mode='rb') as source:
            filename = os.path.join('pseudos', element)
            folder.create_file_from_filelike(source, filename)
            provenance_exclude_list.append(filename)
```
the `provenance_exclude_list` will now for example contain:
```
['pseudos/Si', 'pseudos/Ge']
```
which will now not be stored in the repository. For your use case it will be even easier because you just specify `provenance_exclude_list = ['POSCAR']`` and you're done.

Reply:
Thanks for commenting clearly on this.